File: Frontend/src/app/shared/services/booking.service.ts
import { Injectable } from '@angular/core';
import {Booking} from '../models/booking';
import {Observable} from 'rxjs';
import {RestService} from './rest.service';
import {LocalDate, LocalDateTime} from '@js-joda/core';

@Injectable({
  providedIn: 'root'
})
export class BookingService {

  private endpoint = '/bookings';

  public constructor(private restService: RestService) { }

  public createBooking(body: Booking): Observable<Booking> {
    return this.restService.post(this.endpoint, body) as Observable<Booking>;
  }

  public getOccupationDurationByRoomId(roomId: number): Observable<number> {
    return this.restService.get(this.endpoint + '/get-room-occupation', [{name: 'roomId', value: roomId.toString()}]) as Observable<number>;
  }

  public getRoomAvailability(roomId: number, startBooking: LocalDateTime, endBooking: LocalDateTime): Observable<boolean> {
    return this.restService.get(this.endpoint + '/room-availability',
      [{name: 'roomId', value: roomId.toString()},
        {name: 'startBooking', value: startBooking.toString()},
        {name: 'endBooking', value: endBooking.toString()}]) as Observable<boolean>;
  }

  public findRoomsByDate(date: LocalDate): Observable<Array<Booking>> {
    return this.restService.get(this.endpoint + '/by-date', [{name: 'date', value: date.toString()}]) as Observable<Array<Booking>>;
  }
}

File: Backend/src/main/java/project/raumplaner/RaumplanerApp/booking/service/BookingService.java
package project.raumplaner.RaumplanerApp.booking.service;

import project.raumplaner.RaumplanerApp.base.BaseService;
import project.raumplaner.RaumplanerApp.booking.model.entity.Booking;
import project.raumplaner.RaumplanerApp.room.model.entity.Room;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public interface BookingService extends BaseService<Booking, Long> {

    /**
     * @param date The date of the entity to be found in the database
     * @return All bookings for the certain date with type List<Bookings>
     */
    List<Booking> findByDate(LocalDate date);

    /**
     * @param startBooking The start time of the desired time lapse
     * @param endBooking   The end time of the desired time lapse
     * @return All available rooms for the desired time lapse with type List<Room>
     */
    List<Room> findAvailableRooms(LocalDateTime startBooking, LocalDateTime endBooking);

    /**
     * @param id The id of the room to be checked for occupation time
     * @return Remaining minutes the room is occupied with type Long
     */
    Long calcRoomOccupationInMinutes(Long id);

    /**
     * @param roomId   The id of the room to be checked for occupation time
     * @param baseTime The date and time after which the next booking shall be found to calculate the time difference
     * @return Remaining minutes the room is occupied with type Long
     */
    Long calcTimeToNextBookingByRoom(Long roomId, LocalDateTime baseTime);


    /**
     * @param roomId       The id of the room to be checked for availability in specified time
     * @param startBooking The start date and start time of the availability check
     * @param endBooking   The end date and end time of the availability check
     * @return If the room is available in the specified time period
     */
    Boolean getRoomAvailability(Long roomId, LocalDateTime startBooking, LocalDateTime endBooking);
}
